# 1. Mean Median Mode imputation

    - This techinique is apply when Data is Missing Completely at Random(MCAR)
    - Means there is absolutely no relationship between the data missing and other values.

- Import Required libraries

In [93]:
import pandas as pd
import numpy as np
import seaborn as sns
import plotly 

- load Dataset 

In [94]:
data = pd.read_csv('Titanic.csv')
df1 = data.copy()

In [95]:
df1.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

- Check null values

In [96]:
df1.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

We can see that there is 177 null values in 'Age',687 in 'Cabin' and 2 in 'Embarked' feature.

###  Handle Missing values by using Mean Median Mode imputation

- 1. Replace All NaN values in Age feature with mean

In [97]:
# Calculate Mean
mean = df1['Age'].mean()
mean

29.69911764705882

In [98]:
# Method 1:
# df1['Age']=df1['Age'].fillna(mean) 

In [99]:
# Method 2:
def fill_NAN_Age_mean(df1,variable,mean):
    df1[variable]=df1[variable].fillna(mean)

In [100]:
fill_NAN_Age_mean(df1,'Age',mean)

- Check null values after replace NAN with mean 

In [101]:
df1['Age'].isnull().sum()

0

## -------------------------------------------------------------------------------------------------------------------------------

### Now, Understand maybe there is some outlier are present in Age feature, at that time if we replace NaN with Mean then this feature not follow Gaussion Distribution.

    - So insted of Mean use Median and replace NaN with it

- 2. Replace All NaN values in Age feature with median

In [102]:
df2 = data.copy()

In [103]:
#Calculate Median
median = df2['Age'].median()
median

28.0

In [104]:
# Method 1:
# df2['Age'] = df2['Age'].fillna(df['Age'].median())

In [105]:
#Method 2 :
def fill_NaN_Age_Median(df2,variable,median):
    df2[variable]=df2[variable].fillna(median)

In [106]:
fill_NaN_Age_Median(df2,'Age',median)

In [107]:
df2['Age'].isnull().sum()

0

## -------------------------------------------------------------------------------------------------------------------------------

### If there is some frequent value are present in feature then replace NaN values with that frequent value.
ex. 
    In classs 12th most of student is 17 years old, suppose we have some entries with Missing Age then we assume Age of that student is also 17 years and replace nan with 17 (we assume 17 is a frequent value)

- 3. Replace All NaN values in Age feature with mode

In [108]:
df3 = data.copy()

In [109]:
#Calculate mode
mode=df3['Age'].mode()[0]
mode

24.0

In [110]:
# method 1:
#df3['Age'] = df3['Age'].fillna(mode)

In [111]:
#method 2:
def fill_NAN_Age_mode(df3,variable,mode):
    df3[variable] = df3[variable].fillna(mode)

In [112]:
fill_NAN_Age_mode(df3,'Age',mode)

In [113]:
df3['Age'].isnull().sum()

0

## ------------------------------------------------- End ------------------------------------------------------------